File: pkg/risk/risk.go
package risk

import (
	"math"
	"slices"
)

// VaR computes Value at Risk at the given confidence level.
// Returns the loss threshold that is not exceeded with the specified probability.
// For example, VaR(0.95) returns the 95th percentile of losses (95% confidence).
// losses should contain simulated portfolio losses (positive values represent losses).
// The losses slice is sorted in place during computation.
func VaR(losses []float64, confidence float64) float64 {
	if len(losses) == 0 || confidence <= 0 || confidence >= 1 {
		return 0
	}
	slices.Sort(losses)
	return varSorted(losses, confidence)
}

// CTE computes Conditional Tail Expectation (Expected Shortfall).
// Returns the average of losses exceeding the VaR threshold.
// CTE is more informative than VaR as it captures tail severity.
// The losses slice is sorted in place during computation.
func CTE(losses []float64, confidence float64) float64 {
	if len(losses) == 0 || confidence <= 0 || confidence >= 1 {
		return 0
	}
	slices.Sort(losses)
	return cteSorted(losses, confidence)
}

func varSorted(sorted []float64, confidence float64) float64 {
	idx := int(confidence * float64(len(sorted)-1))
	if idx >= len(sorted) {
		idx = len(sorted) - 1
	}
	return sorted[idx]
}

func cteSorted(sorted []float64, confidence float64) float64 {
	idx := int(confidence * float64(len(sorted)-1))
	if idx >= len(sorted) {
		idx = len(sorted) - 1
	}
	sum := 0.0
	for i := idx; i < len(sorted); i++ {
		sum += sorted[i]
	}
	return sum / float64(len(sorted)-idx)
}

// RiskReport contains comprehensive risk metrics from a simulation.
type RiskReport struct {
	Mean           float64
	StdDev         float64
	Min            float64
	Max            float64
	VaR95          float64
	VaR99          float64
	CTE95          float64
	CTE99          float64
	StdError       float64
	Confidence95Lo float64
	Confidence95Hi float64
}

// ComputeReport generates a full risk report from simulated losses.
// The losses slice is sorted in place during computation.
func ComputeReport(losses []float64) RiskReport {
	n := float64(len(losses))
	if n == 0 {
		return RiskReport{}
	}

	mean := 0.0
	min, max := losses[0], losses[0]
	for _, l := range losses {
		mean += l
		if l < min {
			min = l
		}
		if l > max {
			max = l
		}
	}
	mean /= n

	variance := 0.0
	for _, l := range losses {
		d := l - mean
		variance += d * d
	}
	variance /= n

	slices.Sort(losses)

	stdDev := math.Sqrt(variance)
	stdErr := stdDev / math.Sqrt(n)

	return RiskReport{
		Mean:           mean,
		StdDev:         stdDev,
		StdError:       stdErr,
		Confidence95Lo: mean - 1.96*stdErr,
		Confidence95Hi: mean + 1.96*stdErr,
		Min:            min,
		Max:            max,
		VaR95:          varSorted(losses, 0.95),
		VaR99:          varSorted(losses, 0.99),
		CTE95:          cteSorted(losses, 0.95),
		CTE99:          cteSorted(losses, 0.99),
	}
}

File: zig/src/risk.zig
// risk.zig - Risk measures (VaR, CTE, Expected Shortfall)
//
// Port of Go pkg/risk. Computes risk metrics from simulated portfolio losses.
const std = @import("std");
const math = std.math;
const Allocator = std.mem.Allocator;

/// RiskReport contains comprehensive risk metrics from a simulation.
pub const RiskReport = struct {
    mean: f64 = 0,
    std_dev: f64 = 0,
    min: f64 = 0,
    max: f64 = 0,
    var_95: f64 = 0,
    var_99: f64 = 0,
    cte_95: f64 = 0,
    cte_99: f64 = 0,
    std_error: f64 = 0,
    confidence_95_lo: f64 = 0,
    confidence_95_hi: f64 = 0,
};

/// VaR computes Value at Risk at the given confidence level.
/// Returns the loss threshold that is not exceeded with the specified probability.
/// The losses slice is sorted in place during computation.
pub fn vaR(losses: []f64, confidence: f64) f64 {
    if (losses.len == 0 or confidence <= 0 or confidence >= 1) return 0.0;
    std.sort.pdq(f64, losses, {}, comptime std.sort.asc(f64));
    return varSorted(losses, confidence);
}

/// CTE computes Conditional Tail Expectation (Expected Shortfall).
/// Returns the average of losses exceeding the VaR threshold.
/// The losses slice is sorted in place during computation.
pub fn cte(losses: []f64, confidence: f64) f64 {
    if (losses.len == 0 or confidence <= 0 or confidence >= 1) return 0.0;
    std.sort.pdq(f64, losses, {}, comptime std.sort.asc(f64));
    return cteSorted(losses, confidence);
}

fn varSorted(sorted: []f64, confidence: f64) f64 {
    var idx: usize = @intFromFloat(confidence * @as(f64, @floatFromInt(sorted.len - 1)));
    if (idx >= sorted.len) idx = sorted.len - 1;
    return sorted[idx];
}

fn cteSorted(sorted: []f64, confidence: f64) f64 {
    var idx: usize = @intFromFloat(confidence * @as(f64, @floatFromInt(sorted.len - 1)));
    if (idx >= sorted.len) idx = sorted.len - 1;
    var sum: f64 = 0;
    for (idx..sorted.len) |i| {
        sum += sorted[i];
    }
    return sum / @as(f64, @floatFromInt(sorted.len - idx));
}

/// Generates a full risk report from simulated losses.
/// The losses slice is sorted in place during computation.
pub fn computeReport(losses: []f64) RiskReport {
    const n: f64 = @floatFromInt(losses.len);
    if (n == 0) return .{};

    var mean: f64 = 0;
    var min_val: f64 = losses[0];
    var max_val: f64 = losses[0];
    for (losses) |l| {
        mean += l;
        if (l < min_val) min_val = l;
        if (l > max_val) max_val = l;
    }
    mean /= n;

    var variance: f64 = 0;
    for (losses) |l| {
        const d = l - mean;
        variance += d * d;
    }
    variance /= n;

    std.sort.pdq(f64, losses, {}, comptime std.sort.asc(f64));

    const std_dev = @sqrt(variance);
    const std_err = std_dev / @sqrt(n);

    return .{
        .mean = mean,
        .std_dev = std_dev,
        .min = min_val,
        .max = max_val,
        .var_95 = varSorted(losses, 0.95),
        .var_99 = varSorted(losses, 0.99),
        .cte_95 = cteSorted(losses, 0.95),
        .cte_99 = cteSorted(losses, 0.99),
        .std_error = std_err,
        .confidence_95_lo = mean - 1.96 * std_err,
        .confidence_95_hi = mean + 1.96 * std_err,
    };
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

test "VaR basic" {
    const testing = std.testing;
    var losses = [_]f64{ 10, 20, 30, 40, 50, 60, 70, 80, 90, 100 };
    const result = vaR(&losses, 0.95);
    try testing.expect(result >= 90);
    try testing.expect(result <= 100);
}

test "CTE basic" {
    const testing = std.testing;
    var losses = [_]f64{ 10, 20, 30, 40, 50, 60, 70, 80, 90, 100 };
    const result = cte(&losses, 0.95);
    try testing.expect(result >= 90);
    try testing.expect(result <= 100);
}

test "CTE >= VaR" {
    const testing = std.testing;
    var losses1 = [_]f64{ 10, 20, 30, 40, 50, 60, 70, 80, 90, 100 };
    var losses2 = [_]f64{ 10, 20, 30, 40, 50, 60, 70, 80, 90, 100 };
    const var_result = vaR(&losses1, 0.95);
    const cte_result = cte(&losses2, 0.95);
    try testing.expect(cte_result >= var_result);
}

test "ComputeReport" {
    const testing = std.testing;
    var losses = [_]f64{ 10, 20, 30, 40, 50, 60, 70, 80, 90, 100 };
    const report = computeReport(&losses);

    try testing.expectApproxEqAbs(@as(f64, 55), report.mean, 1e-10);
    try testing.expectApproxEqAbs(@as(f64, 10), report.min, 1e-10);
    try testing.expectApproxEqAbs(@as(f64, 100), report.max, 1e-10);
    try testing.expect(report.std_dev > 0);
    try testing.expect(report.var_95 > 0);
    try testing.expect(report.cte_95 >= report.var_95);
}

test "ComputeReport empty" {
    const testing = std.testing;
    var losses: [0]f64 = undefined;
    const report = computeReport(&losses);
    try testing.expectApproxEqAbs(@as(f64, 0), report.mean, 1e-10);
}
